File: NetFramework/XHtmlEditor/CS/C1EditorQuickStart/ToolStrips/ToolStripMain.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;
using C1.Win.C1SpellChecker;

namespace C1EditorQuickStart.ToolStrips
{
    class ToolStripMain : ToolStripBase
    {
        //----------------------------------------------------------------------------
        #region ** fields

        string _fileName;
        bool _dirty;
        C1SpellChecker _spellChecker;
        ToolStripButton _btnNew, _btnOpen, _btnSave, _btnPrint;
        ToolStripButton _btnCut, _btnCopy , _btnPaste;
        ToolStripButton _btnUndo, _btnRedo;
        ToolStripButton _btnFind, _btnReplace;
        ToolStripButton _btnShowErrors, _btnSpell;
        ToolStripButton _btnWordCount;
        ToolStripButton _btnHelp;

        #endregion

        //----------------------------------------------------------------------------
        #region ** object model

        public C1SpellChecker SpellChecker
        {
            get { return _spellChecker; }
            set { _spellChecker = value; }
        }

        #endregion

        //----------------------------------------------------------------------------
        #region ** event handlers

        // initialize the toolstrip
        protected override void OnInitialize()
        {
            // populate toolstrip
            _btnNew = AddButton("New", Properties.Resources.New_small, New_Click);
            _btnOpen = AddButton("Open", Properties.Resources.Open_small, Open_Click);
            _btnSave = AddButton("Save", Properties.Resources.Save_small, Save_Click);
            _btnPrint = AddButton("Print", Properties.Resources.Print_small, Print_Click);
            Items.Add(new ToolStripSeparator());
            _btnCut = AddButton("Cut", Properties.Resources.Cut_small, Cut_Click);
            _btnCopy = AddButton("Copy", Properties.Resources.Copy_small, Copy_Click);
            _btnPaste = AddButton("Paste", Properties.Resources.Paste_small, Paste_Click);
            Items.Add(new ToolStripSeparator());
            _btnUndo = AddButton("Undo", Properties.Resources.Undo_small, Undo_Click);
            _btnRedo = AddButton("Redo", Properties.Resources.Redo_small, Redo_Click);
            Items.Add(new ToolStripSeparator());
            _btnShowErrors = AddButton("Show Spelling Errors", Properties.Resources.ShowErrors_small, ShowSpellingErrors_Click);
            _btnSpell = AddButton("Spell", Properties.Resources.Spelling_small, Spell_Click);
            Items.Add(new ToolStripSeparator());
            _btnFind = AddButton("Find", Properties.Resources.Find_small, Find_Click);
            _btnReplace = AddButton("Replace", Properties.Resources.Replace_small, Replace_Click);
            Items.Add(new ToolStripSeparator());
            _btnWordCount = AddButton("Word Count", Properties.Resources.WordCount_small, WordCount_Click);
            Items.Add(new ToolStripSeparator());
            _btnHelp = AddButton("Help", Properties.Resources.Help_small, Help_Click);
        }

        // enable/disable buttons
        public override void UpdateState()
        {
            Enabled = Editor != null;
            _btnCopy.Enabled = _btnCut.Enabled = Editor.CanCopy;
            _btnPaste.Enabled = Editor.CanPasteAsText; // CanPaste
            _btnUndo.Enabled = Editor.CanUndo();
            _btnRedo.Enabled = Editor.CanRedo();
            bool isDesign = Editor.Mode == C1.Win.C1Editor.EditorMode.Design;
            bool isPreview = Editor.Mode == C1.Win.C1Editor.EditorMode.Preview;
            _btnShowErrors.Enabled = isDesign && SpellChecker != null;
            _btnSpell.Enabled = isDesign && SpellChecker != null;
            _btnFind.Enabled = !isPreview;
            _btnReplace.Enabled = !isPreview;
        }

        // keep track of changes in the document
        protected override void OnDocumentChanged(EventArgs e)
        {
            SetDirty(true);
        }
        
        // keep track of changes in the document
        protected override void OnModeChanged(EventArgs e)
        {
            UpdateState();
        }

        protected override void OnEditorKeyDownHandler(KeyEventArgs e)
        {
            if (e.Control && e.KeyCode == Keys.S)
            {
                e.Handled = true;
                Save_Click(null, EventArgs.Empty);
            }
            else if (e.Control && e.KeyCode == Keys.O)
            {
                e.Handled = true;
                Open_Click(null, EventArgs.Empty);
            }
            else if (e.Control && e.KeyCode == Keys.P)
            {
                e.Handled = true;
                Print_Click(null, EventArgs.Empty);
            }
            else if (e.Control && e.KeyCode == Keys.N)
            {
                e.Handled = true;
                New_Click(null, EventArgs.Empty);
            }
        }

        // file commands
        void New_Click(object sender, EventArgs e)
        {
            NewDocument();
        }
        void Open_Click(object sender, EventArgs e)
        {
            LoadDocument();
        }
        void Save_Click(object sender, EventArgs e)
        {
            SaveDocument();
        }

        // preview/print
        void Print_Click(object sender, EventArgs e)
        {
            Editor.PrintPreview();
        }

        // clipboard
        void Cut_Click(object sender, EventArgs e)
        {
            Editor.Cut();
        }
        void Copy_Click(object sender, EventArgs e)
        {
            Editor.Copy();
        }
        void Paste_Click(object sender, EventArgs e)
        {
            Editor.PasteAsText();//.Paste();
        }

        // undo/redo
        void Undo_Click(object sender, EventArgs e)
        {
            Editor.Undo();
        }
        void Redo_Click(object sender, EventArgs e)
        {
            Editor.Redo();
        }

        // find/replace
        void Find_Click(object sender, EventArgs e)
        {
            Editor.ShowDialog(C1.Win.C1Editor.DialogType.Find);
        }
        void Replace_Click(object sender, EventArgs e)
        {
            Editor.ShowDialog(C1.Win.C1Editor.DialogType.Replace);
        }

        // spelling
        void ShowSpellingErrors_Click(object sender, EventArgs e)
        {
            if (SpellChecker != null)
            {
                bool show = !_btnShowErrors.Checked;
                _btnShowErrors.Checked = show;
                SpellChecker.SetActiveSpellChecking(Editor, Editor.GetActiveXInstance(), show);
            }
        }
        void Spell_Click(object sender, EventArgs e)
        {
            if (SpellChecker != null)
            {
                SpellChecker.CheckControl(Editor, Editor.GetActiveXInstance());
            }
        }

        // word count
        void WordCount_Click(object sender, EventArgs e)
        {
            string text = Editor.Text;
            int chars = 0;
            int nonSpaceChars = 0;
            int words = 0;
            int paragraphs = 0;
            int sentences = 0;

            int length = text.Length;

            // calculate statistics
            for (int i = 0; i < length; i++)
            {
                char c = text[i];

                // count chars
                chars++;

                // count non-space chars
                if (!char.IsWhiteSpace(c))
                {
                    nonSpaceChars++;
                }

                // count paragraphs
                if (c == '\n' || i == length - 1)
                {
                    if (i == length - 1 || text[i + 1] != '\n')
                    {
                        paragraphs++;
                    }
                }

                // count sentences
                if (c == '.' || c == '!' || c == '?' || i == length - 1)
                {
                    if (i == length - 1 || char.IsWhiteSpace(text, i + 1))
                    {
                        sentences++;
                    }
                }

                // count words
                if (char.IsLetterOrDigit(c))
                {
                    if (i == length - 1 || !char.IsLetterOrDigit(text, i + 1))
                    {
                        words++;
                    }
                }
            }

            // show statistics
            string msg = string.Format(
                "Words: {0:n0}\r\n" +
                "Characters: {1:n0}\r\n" +
                "Non-Space Characters: {2:n0}\r\n" +
                "Sentences: {3:n0}\r\n" +
                "Paragraphs: {4:n0}\r\n" +
                "Average Word Length: {5:n1}\r\n" +
                "Average Sentence Length: {6:n1}\r\n" +
                "Average Paragraph Length: {7:n1}\r\n",
                words, chars, nonSpaceChars, sentences, paragraphs,
                words > 0 ? nonSpaceChars / (float)words : 0f,
                sentences > 0 ? nonSpaceChars / (float)sentences : 0f,
                paragraphs > 0 ? nonSpaceChars / (float)paragraphs : 0f);
            MessageBox.Show(msg, "Word Count");
        }

        // help
        void Help_Click(object sender, EventArgs e)
        {
            string msg = string.Format(
                Properties.Resources.About,
                Editor.GetType().Assembly.GetName().Version);
            MessageBox.Show(msg, ProductName);
        }

        #endregion

        //----------------------------------------------------------------------------
        #region ** implementation

        // keep track of changes
        void SetDirty(bool dirty)
        {
            _dirty = dirty;
            UpdateCaption();
        }

        // check with user before discarding changes
        bool OKToDiscardChanges()
        {
            if (_dirty)
            {
                DialogResult dr = MessageBox.Show(
                    Properties.Resources.PromptSaveChanges,
                    ProductName,
                    MessageBoxButtons.YesNoCancel,
                    MessageBoxIcon.Question);

                switch (dr)
                {
                    // save document, then proceed
                    case DialogResult.Yes:
                        return SaveDocument();

                    // proceed without saving
                    case DialogResult.No:
                        return true;

                    // do not proceed
                    case DialogResult.Cancel:
                        return false;
                }
            }

            // no changes, ok to proceed
            return true;
        }

        // show current document and modified state
        void UpdateCaption()
        {
            Form form = TopLevelControl as Form;
            if (form != null)
            {
                StringBuilder sb = new StringBuilder(ProductName);
                if (!string.IsNullOrEmpty(_fileName))
                {
                    sb.AppendFormat(" - {0}", Path.GetFileName(_fileName));
                    if (_dirty)
                    {
                        sb.Append(" *");
                    }
                }
                form.Text = sb.ToString();
            }
        }

        // create a new document
        bool NewDocument()
        {
            if (OKToDiscardChanges())
            {
                Editor.Document = new System.Xml.XmlDocument();
                _fileName = null;
                SetDirty(false);
                return true;
            }
            return false;
        }

        // load a document into the editor
        bool LoadDocument()
        {
            if (OKToDiscardChanges())
            {
                using (OpenFileDialog dlg = new OpenFileDialog())
                {
                    // get file name
                    dlg.Filter = Properties.Resources.FileFilter;
                    dlg.DefaultExt = Properties.Resources.DefaultExt;
                    if (dlg.ShowDialog() == DialogResult.OK)
                    {
                        try
                        {
                            // load document
                            Editor.LoadXml(dlg.FileName);
                            _fileName = dlg.FileName;
                            SetDirty(false);
                            return true;
                        }
                        catch (Exception x)
                        {
                            MessageBox.Show(x.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                            return false;
                        }
                    }
                }
            }

            // canceled...
            return false;
        }

        // save current document using current file name
        bool SaveDocument()
        {
            // no name? go get one...
            if (string.IsNullOrEmpty(_fileName))
            {
                return SaveDocumentAs();
            }

            // got the name, save the file
            try
            {
                Editor.SaveXml(_fileName);
                SetDirty(false);
                return true;
            }
            catch (Exception x)
            {
                MessageBox.Show(x.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return false;
            }
        }

        // save current document using a new name
        bool SaveDocumentAs()
        {
            using (SaveFileDialog dlg = new SaveFileDialog())
            {
                dlg.Filter = Properties.Resources.FileFilter;
                dlg.DefaultExt = Properties.Resources.DefaultExt;
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    _fileName = dlg.FileName;
                    return SaveDocument();
                }
            }
            return false;
        }

        #endregion

    }
}

File: NetFramework/XHtmlEditor/VB/C1EditorQuickStart/ToolStrips/ToolStripMain.vb
Imports System.IO
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Text
Imports System.Windows.Forms
Imports C1.Win.SpellChecker

Namespace ToolStrips
	Class ToolStripMain
		Inherits ToolStripBase
		'----------------------------------------------------------------------------
		#Region "** fields"

		Private _fileName As String
		Private _dirty As Boolean
		Private _spellChecker As C1SpellChecker
		Private _btnNew As ToolStripButton, _btnOpen As ToolStripButton, _btnSave As ToolStripButton, _btnPrint As ToolStripButton
		Private _btnCut As ToolStripButton, _btnCopy As ToolStripButton, _btnPaste As ToolStripButton
		Private _btnUndo As ToolStripButton, _btnRedo As ToolStripButton
		Private _btnFind As ToolStripButton, _btnReplace As ToolStripButton
		Private _btnShowErrors As ToolStripButton, _btnSpell As ToolStripButton
		Private _btnWordCount As ToolStripButton
		Private _btnHelp As ToolStripButton

		#End Region

		'----------------------------------------------------------------------------
		#Region "** object model"

		Public Property SpellChecker() As C1SpellChecker
			Get
				Return _spellChecker
			End Get
			Set
				_spellChecker = value
			End Set
		End Property

		#End Region

		'----------------------------------------------------------------------------
		#Region "** event handlers"

		' initialize the toolstrip
		Protected Overloads Overrides Sub OnInitialize()
			' populate toolstrip
            _btnNew = AddButton("New", Resources.New_small, AddressOf New_Click)
            _btnOpen = AddButton("Open", Resources.Open_small, AddressOf Open_Click)
            _btnSave = AddButton("Save", Resources.Save_small, AddressOf Save_Click)
            _btnPrint = AddButton("Print", Resources.Print_small, AddressOf Print_Click)
            Items.Add(New ToolStripSeparator())
            _btnCut = AddButton("Cut", Resources.Cut_small, AddressOf Cut_Click)
            _btnCopy = AddButton("Copy", Resources.Copy_small, AddressOf Copy_Click)
            _btnPaste = AddButton("Paste", Resources.Paste_small, AddressOf Paste_Click)
            Items.Add(New ToolStripSeparator())
            _btnUndo = AddButton("Undo", Resources.Undo_small, AddressOf Undo_Click)
            _btnRedo = AddButton("Redo", Resources.Redo_small, AddressOf Redo_Click)
            Items.Add(New ToolStripSeparator())
            _btnShowErrors = AddButton("Show Spelling Errors", Resources.ShowErrors_small, AddressOf ShowSpellingErrors_Click)
            _btnSpell = AddButton("Spell", Resources.Spelling_small, AddressOf Spell_Click)
            Items.Add(New ToolStripSeparator())
            _btnFind = AddButton("Find", Resources.Find_small, AddressOf Find_Click)
            _btnReplace = AddButton("Replace", Resources.Replace_small, AddressOf Replace_Click)
            Items.Add(New ToolStripSeparator())
            _btnWordCount = AddButton("Word Count", Resources.WordCount_small, AddressOf WordCount_Click)
            Items.Add(New ToolStripSeparator())
            _btnHelp = AddButton("Help", Resources.Help_small, AddressOf Help_Click)
        End Sub

        ' enable/disable buttons
        Public Overloads Overrides Sub UpdateState()
            Dim isDesign = (Editor.Mode = C1.Win.C1Editor.EditorMode.Design)
            Dim isPreview = (Editor.Mode = C1.Win.C1Editor.EditorMode.Preview)

            Enabled = Editor IsNot Nothing
            _btnCopy.Enabled = InlineAssignHelper(_btnCut.Enabled, Editor.CanCopy)
            _btnPaste.Enabled = Editor.CanPasteAsText
            ' CanPaste
            _btnUndo.Enabled = Editor.CanUndo()
            _btnRedo.Enabled = Editor.CanRedo()

            _btnShowErrors.Enabled = isDesign And (SpellChecker IsNot Nothing)
            _btnSpell.Enabled = isDesign And (SpellChecker IsNot Nothing)
            _btnFind.Enabled = Not isPreview
            _btnReplace.Enabled = Not isPreview
        End Sub

        ' keep track of changes in the document
        Protected Overloads Overrides Sub OnDocumentChanged(ByVal e As EventArgs)
            SetDirty(True)
        End Sub

        ' keep track of changes in the document
        Protected Overloads Overrides Sub OnModeChanged(ByVal e As EventArgs)
            UpdateState()
        End Sub

        Protected Overloads Overrides Sub OnEditorKeyDownHandler(ByVal e As KeyEventArgs)
            If e.Control And e.KeyCode = Keys.S Then
                e.Handled = True
                Save_Click(Nothing, e)
            ElseIf e.Control And e.KeyCode = Keys.O Then
                e.Handled = True
                Open_Click(Nothing, e)
            ElseIf e.Control And e.KeyCode = Keys.P Then
                e.Handled = True
                Print_Click(Nothing, e)
            ElseIf e.Control And e.KeyCode = Keys.N Then
                e.Handled = True
                New_Click(Nothing, e)
            End If
        End Sub

        ' file commands
        Private Sub New_Click(ByVal sender As Object, ByVal e As EventArgs)
            NewDocument()
        End Sub
        Private Sub Open_Click(ByVal sender As Object, ByVal e As EventArgs)
            LoadDocument()
        End Sub
        Private Sub Save_Click(ByVal sender As Object, ByVal e As EventArgs)
            SaveDocument()
        End Sub

        ' preview/print
        Private Sub Print_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.PrintPreview()
        End Sub

        ' clipboard
        Private Sub Cut_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.Cut()
        End Sub
        Private Sub Copy_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.Copy()
        End Sub
        Private Sub Paste_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.PasteAsText()
            '.Paste();
        End Sub

        ' undo/redo
        Private Sub Undo_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.Undo()
        End Sub
        Private Sub Redo_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.Redo()
        End Sub

        ' find/replace
        Private Sub Find_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.ShowDialog(C1.Win.C1Editor.DialogType.Find)
        End Sub
        Private Sub Replace_Click(ByVal sender As Object, ByVal e As EventArgs)
            Editor.ShowDialog(C1.Win.C1Editor.DialogType.Replace)
        End Sub

        ' spelling
        Private Sub ShowSpellingErrors_Click(ByVal sender As Object, ByVal e As EventArgs)
            If SpellChecker IsNot Nothing Then
                Dim show As Boolean = Not _btnShowErrors.Checked
                _btnShowErrors.Checked = show
                SpellChecker.SetActiveSpellChecking(Editor, Editor.GetActiveXInstance(), show)
            End If
        End Sub
        Private Sub Spell_Click(ByVal sender As Object, ByVal e As EventArgs)
            If SpellChecker IsNot Nothing Then
                SpellChecker.CheckControl(Editor, Editor.GetActiveXInstance())
            End If
        End Sub

        ' word count
        Private Sub WordCount_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim text As String = Editor.Text
            Dim chars As Integer = 0
            Dim nonSpaceChars As Integer = 0
            Dim words As Integer = 0
            Dim paragraphs As Integer = 0
            Dim sentences As Integer = 0

            Dim length As Integer = text.Length

            ' calculate statistics
            For i As Integer = 0 To length - 1
                Dim c As Char = text(i)

                ' count chars
                chars += 1

                ' count non-space chars
                If Not Char.IsWhiteSpace(c) Then
                    nonSpaceChars += 1
                End If

                ' count paragraphs
                If c = ControlChars.Lf OrElse i = length - 1 Then
                    If i = length - 1 OrElse text(i + 1) <> ControlChars.Lf Then
                        paragraphs += 1
                    End If
                End If

                ' count sentences
                If c = "."c OrElse c = "!"c OrElse c = "?"c OrElse i = length - 1 Then
                    If i = length - 1 OrElse Char.IsWhiteSpace(text, i + 1) Then
                        sentences += 1
                    End If
                End If

                ' count words
                If Char.IsLetterOrDigit(c) Then
                    If i = length - 1 OrElse Not Char.IsLetterOrDigit(text, i + 1) Then
                        words += 1
                    End If
                End If
            Next

            ' show statistics
            Dim nonSpaceChars_words As Single = 0.0F
            Dim nonSpaceChars_paragraphs As Single = 0.0F
            Dim nonSpaceChars_sentences As Single = 0.0F
            If words > 0 Then
                nonSpaceChars_words = nonSpaceChars / CSng(words)
            End If

            If sentences > 0 Then
                nonSpaceChars_sentences = nonSpaceChars / CSng(sentences)
            End If

            If paragraphs > 0 Then
                nonSpaceChars_paragraphs = nonSpaceChars / CSng(paragraphs)
            End If

            Dim msg As String = String.Format("Words: {0:n0}" & vbCr & vbLf & "Characters: {1:n0}" & vbCr & vbLf & "Non-Space Characters: {2:n0}" & vbCr & vbLf & "Sentences: {3:n0}" & vbCr & vbLf & "Paragraphs: {4:n0}" & vbCr & vbLf & "Average Word Length: {5:n1}" & vbCr & vbLf & "Average Sentence Length: {6:n1}" & vbCr & vbLf & "Average Paragraph Length: {7:n1}" & vbCr & vbLf, words, chars, nonSpaceChars, sentences, paragraphs, _
             nonSpaceChars_words, nonSpaceChars_sentences, nonSpaceChars_paragraphs)
            MessageBox.Show(msg, "Word Count")
        End Sub

        ' help
        Private Sub Help_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim msg As String = String.Format(Resources.About, Editor.[GetType]().Assembly.GetName().Version)
            MessageBox.Show(msg, ProductName)
        End Sub

#End Region

        '----------------------------------------------------------------------------
#Region "** implementation"

        ' keep track of changes
        Private Sub SetDirty(ByVal dirty As Boolean)
            _dirty = dirty
            UpdateCaption()
        End Sub

        ' check with user before discarding changes
        Private Function OKToDiscardChanges() As Boolean
            If _dirty Then
                Dim dr As DialogResult = MessageBox.Show(Resources.PromptSaveChanges, ProductName, MessageBoxButtons.YesNoCancel, MessageBoxIcon.Question)

                Select Case dr
                    ' save document, then proceed
                    Case DialogResult.Yes
                        Return SaveDocument()

                        ' proceed without saving
                    Case DialogResult.No
                        Return True

                        ' do not proceed
                    Case DialogResult.Cancel
                        Return False
                End Select
            End If

            ' no changes, ok to proceed
            Return True
        End Function

        ' show current document and modified state
        Private Sub UpdateCaption()
            Dim form As Form = TryCast(TopLevelControl, Form)
            If form IsNot Nothing Then
                Dim sb As StringBuilder = New StringBuilder(ProductName)
                If Not String.IsNullOrEmpty(_fileName) Then
                    sb.AppendFormat(" - {0}", Path.GetFileName(_fileName))
                    If _dirty Then
                        sb.Append(" *")
                    End If
                End If
                form.Text = sb.ToString()
            End If
        End Sub

        ' create a new document
        Private Function NewDocument() As Boolean
            If OKToDiscardChanges() Then
                Editor.Document = New System.Xml.XmlDocument()
                _fileName = Nothing
                SetDirty(False)
                Return True
            End If
            Return False
        End Function

        ' load a document into the editor
        Private Function LoadDocument() As Boolean
            If OKToDiscardChanges() Then
                Using dlg As OpenFileDialog = New OpenFileDialog()
                    ' get file name
                    dlg.Filter = Resources.FileFilter
                    dlg.DefaultExt = Resources.DefaultExt
                    If dlg.ShowDialog() = DialogResult.OK Then
                        Try
                            ' load document
                            Editor.LoadXml(dlg.FileName)
                            _fileName = dlg.FileName
                            SetDirty(False)
                            Return True
                        Catch x As Exception
                            MessageBox.Show(x.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
                            Return False
                        End Try
                    End If
                End Using
            End If

            ' canceled...
            Return False
        End Function

        ' save current document using current file name
        Private Function SaveDocument() As Boolean
            ' no name? go get one...
            If String.IsNullOrEmpty(_fileName) Then
                Return SaveDocumentAs()
            End If

            ' got the name, save the file
            Try
                Editor.SaveXml(_fileName)
                SetDirty(False)
                Return True
            Catch x As Exception
                MessageBox.Show(x.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.[Error])
                Return False
            End Try
        End Function

        ' save current document using a new name
        Private Function SaveDocumentAs() As Boolean
            Using dlg As SaveFileDialog = New SaveFileDialog()
                dlg.Filter = Resources.FileFilter
                dlg.DefaultExt = Resources.DefaultExt
                If dlg.ShowDialog() = DialogResult.OK Then
                    _fileName = dlg.FileName
                    Return SaveDocument()
                End If
            End Using
            Return False
        End Function
        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, ByVal value As T) As T
            target = value
            Return value
        End Function

#End Region

	End Class
End Namespace
